core/drivers/crypto/hisilicon/include/hisi_qm.h

```diff
@@ -0,0 +1,274 @@
+	uint16_t pasid;
+	/*
+	 * rsv : 5
+	 * head_sig : 1
+	 * tail_sig : 1
+	 * pasid_en : 1
+	 * rsv : 8
+	 */
+	uint16_t w11;
+	uint16_t cq_num;
+	/*
+	 * priority(Credit): 4
+	 * order(order/fc/close/rsv) : 4
+	 * type : 4
+	 * rsv : 4
+	 */
+	uint16_t w13;
+	uint32_t rsv1;
+};
+
+struct qm_cqc {
+	uint16_t head;
+	uint16_t tail;
+	uint32_t base_l;
+	uint32_t base_h;
+	/*
+	 * qes : 12
+	 * cqe_size : 4
+	 * rsv(stash_nid/stash_en) : 16
+	 */
+	uint32_t dw3;
+	uint16_t rand_data;
+	uint16_t rsv0;
+	uint16_t pasid;
+	/*
+	 * pasid_en : 1
+	 * rsv : 4
+	 * head_sig : 1
+	 * tail_sig : 1
+	 * rsv : 9
+	 */
+	uint16_t w11;
+	/*
+	 * phase : 1
+	 * c_flag : 1
+	 * stash_vld : 1
+	 */
+	uint32_t dw6;
+	uint32_t rsv1;
+};
+
+struct qm_cqe {
+	uint32_t rsv0;
+	uint16_t cmd_id;
+	uint16_t rsv1;
+	uint16_t sq_head;
+	uint16_t sq_id;
+	uint16_t rsv2;
+	/*
+	 * p : 1
+	 * status : 15
+	 */
+	uint16_t w7;
+};
+
+struct hisi_qp {
+	struct hisi_qm *qm;
+	uint32_t qp_id;
+	uint8_t sq_type;
+	uint16_t sq_tail;
+	uint16_t cq_head;
+	bool cqc_phase;
+
+	void *sqe;
+	struct qm_cqe *cqe;
+	paddr_t sqe_dma;
+	paddr_t cqe_dma;
+
+	int32_t (*fill_sqe)(void *sqe, void *msg);
+	int32_t (*parse_sqe)(void *sqe, void *msg);
+};
+
+struct hisi_qm {
+	enum qm_fun_type fun_type;
+	vaddr_t io_base;
+	uint32_t io_size;
+	uint32_t vfs_num;
+	uint32_t version;
+
+	struct qm_sqc *sqc;
+	struct qm_cqc *cqc;
+	paddr_t sqc_dma;
+	paddr_t cqc_dma;
+	uint32_t sqe_size;
+	uint32_t sqe_log2_size;
+	uint32_t qp_base;
+	uint32_t qp_num;
+	uint32_t qp_in_used;
+	uint32_t qp_idx;
+	struct hisi_qp *qp_array;
+	struct mutex qp_lock; /* protect the qp instance */
+
+	int32_t (*dev_status_check)(struct hisi_qm *qm);
+};
+
+enum hisi_drv_status {
+	DRVCRYPT_NO_ERR = 0,
+	DRVCRYPT_FAIL = 1,
+	DRVCRYPT_EIO = 5,
+	DRVCRYPT_EAGAIN = 11,
+	DRVCRYPT_ENOMEM = 12,
+	DRVCRYPT_EFAULT = 14,
+	DRVCRYPT_EBUSY = 16,
+	DRVCRYPT_ENODEV = 19,
+	DRVCRYPT_EINVAL = 22,
+	DRVCRYPT_ETMOUT = 110,
+	DRVCRYPT_ENOPROC,
+	DRVCRYPT_IN_EPARA,
+	DRVCRYPT_VERIFY_ERR,
+	DRVCRYPT_HW_EACCESS,
+};
+
+#define readl_relaxed_poll_timeout(addr, val, cond, delay_us, timeout_us) \
+({ \
+	uint64_t timeout = 0; \
+	uint64_t _delay_us = delay_us; \
+	bool flag = false; \
+	while (timeout < (timeout_us)) { \
+		(val) = io_read32(addr); \
+		if (cond) { \
+			flag = true; \
+			break; \
+		} \
+		timeout += (_delay_us); \
+		udelay(_delay_us); \
+	} \
+	(flag) ? 0 : -DRVCRYPT_ETMOUT; \
+})
+
+struct acc_device {
+	struct hisi_qm qm;
+	vaddr_t io_base;
+	uint32_t io_size;
+	uint32_t vfs_num;
+	uint32_t endian;
+	enum qm_fun_type fun_type;
+	SLIST_ENTRY(acc_device) link;
+};
+
+/**
+ *@Description: Get the version information of QM hardware
+ *@param qm: Handle of Queue Management module
+ */
+void hisi_qm_get_version(struct hisi_qm *qm);
+
+/**
+ *@Description: Init QM for Kunpeng drv
+ *@param qm: Handle of Queue Management module
+ *@return success: 0，fail: -DRVCRYPT_EBUSY/DRVCRYPT_EINVAL
+ */
+int32_t hisi_qm_init(struct hisi_qm *qm);
```

Comment:
Why not returning type `enum hisi_drv_status`. It would make the implementation better and prevent one messes up between several kinds of return value enums. This comment applies to all `hsis_xxx()` function returning these error codes.

Note I don't think you need a negative value. Testing a non-zero retrun code would be sufficient to detect an error condition.